compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala

```diff
@@ -51,53 +51,53 @@ class ArrayApplyOptTest extends DottyBytecodeTest {
   @Test def testArrayApplyBoolean = {
     val init = List(Op(DUP), Op(ICONST_0), Op(ICONST_1), Op(BASTORE), Op(DUP), Op(ICONST_1), Op(ICONST_0), Op(BASTORE))
     test("Array(true, false)", newArray2Opcodes(T_BOOLEAN, init))
-    test("IArray(true, false)", newArray2Opcodes(T_BOOLEAN, init :+ TypeOp(CHECKCAST, "[Z")))
+    test("IArray(true, false)", newArray2Opcodes(T_BOOLEAN, init))
   }
 
   @Test def testArrayApplyByte = {
     val init = List(Op(DUP), Op(ICONST_0), Op(ICONST_1), Op(BASTORE), Op(DUP), Op(ICONST_1), Op(ICONST_2), Op(BASTORE))
-    test("Array[Byte](1, 2)", newArray2Opcodes(T_BYTE, init))
-    test("IArray[Byte](1, 2)", newArray2Opcodes(T_BYTE, init :+ TypeOp(CHECKCAST, "[B")))
+    test("Array(1: Byte, 2: Byte)", newArray2Opcodes(T_BYTE, init))
+    test("IArray(1: Byte, 2: Byte)", newArray2Opcodes(T_BYTE, init))
   }
 
   @Test def testArrayApplyShort = {
     val init = List(Op(DUP), Op(ICONST_0), Op(ICONST_1), Op(SASTORE), Op(DUP), Op(ICONST_1), Op(ICONST_2), Op(SASTORE))
-    test("Array[Short](1, 2)", newArray2Opcodes(T_SHORT, init))
-    test("IArray[Short](1, 2)", newArray2Opcodes(T_SHORT, init :+ TypeOp(CHECKCAST, "[S")))
+    test("Array(1: Short, 2: Short)", newArray2Opcodes(T_SHORT, init))
+    test("IArray(1: Short, 2: Short)", newArray2Opcodes(T_SHORT, init))
   }
 
   @Test def testArrayApplyInt = {
     val init = List(Op(DUP), Op(ICONST_0), Op(ICONST_1), Op(IASTORE), Op(DUP), Op(ICONST_1), Op(ICONST_2), Op(IASTORE))
     test("Array(1, 2)", newArray2Opcodes(T_INT, init))
-    test("IArray(1, 2)", newArray2Opcodes(T_INT, init :+ TypeOp(CHECKCAST, "[I")))
+    test("IArray(1, 2)", newArray2Opcodes(T_INT, init))
 
-    val init2 = List(Op(DUP), Op(ICONST_0), Field(GETSTATIC, "Foo$", "MODULE$", "LFoo$;"), Invoke(INVOKEVIRTUAL, "Foo$", "t", "()I", false), Op(IASTORE), Op(DUP), Op(ICONST_1), Field(GETSTATIC, "Foo$", "MODULE$", "LFoo$;"), Invoke(INVOKEVIRTUAL, "Foo$", "t", "()I", false), Op(IASTORE))
-    test("""Array[T](t, t)""", newArray2Opcodes(T_INT, init2))
-    test("""IArray[T](t, t)""", newArray2Opcodes(T_INT, init2 :+ TypeOp(CHECKCAST, "[I")))
+    val init2 = List(Op(DUP), Op(ICONST_0), Field(GETSTATIC, "Foo$", "MODULE$", "LFoo$;"), Invoke(INVOKEVIRTUAL, "Foo$", "t", "()I", false), Op(IASTORE), Op(DUP), Op(ICONST_1), Field(GETSTATIC, "Foo$", "MODULE$", "LFoo$;"), Invoke(INVOKEVIRTUAL, "Foo$", "t", "()I", false), Op(IASTORE), TypeOp(CHECKCAST, "[I"))
```

Comment:
Now we have a cast in both branches. Aren't the type ascriptions not supposed to avoid this extra cast in bytecode?

Reply:
The erased result type of the generic Array.apply is `Object`, so a checkcast is necessary to actually use it as an Array. Without the type ascription, the compiler doesn't always insert that checkcast because we don't actually do anything with the Array.